examples/heal/local-nsmgr-restart/README.md

```diff
@@ -0,0 +1,156 @@
+
+Make sure that you have completed steps from [basic](../../basic) or [memory](../../memory) setup.
+
+## Run
+
+Create test namespace:
+```bash
+NAMESPACE=($(kubectl create -f ../namespace.yaml)[0])
+NAMESPACE=${NAMESPACE:10}
+```
+
+Register namespace in `spire` server:
+```bash
+kubectl exec -n spire spire-server-0 -- \
+/opt/spire/bin/spire-server entry create \
+-spiffeID spiffe://example.org/ns/${NAMESPACE}/sa/default \
+-parentID spiffe://example.org/ns/spire/sa/spire-agent \
+-selector k8s:ns:${NAMESPACE} \
+-selector k8s:sa:default
+```
+
+Get nodes exclude control-plane:
+```bash
+NODES=($(kubectl get nodes -o go-template='{{range .items}}{{ if not .spec.taints  }}{{index .metadata.labels "kubernetes.io/hostname"}} {{end}}{{end}}'))
+```
+
+Create customization file:
+```bash
+cat > kustomization.yaml <<EOF
+---
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+
+namespace: ${NAMESPACE}
+
+bases:
+- ../../../apps/nsc-kernel
+- ../../../apps/nse-kernel
+
+patchesStrategicMerge:
+- patch-nsc.yaml
+- patch-nse.yaml
+EOF
+```
+
+Create NSC patch:
+```bash
+cat > patch-nsc.yaml <<EOF
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nsc-kernel
+spec:
+  template:
+    spec:
+      containers:
+        - name: nsc
+          env:
+            - name: NSM_NETWORK_SERVICES
+              value: kernel://icmp-responder/nsm-1
+
+      nodeSelector:
+        kubernetes.io/hostname: ${NODES[0]}
+EOF
+
+```
+Create NSE patch:
+```bash
+cat > patch-nse.yaml <<EOF
+---
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: nse-kernel
+spec:
+  template:
+    spec:
+      containers:
+        - name: nse
+          env:
+            - name: NSE_CIDR_PREFIX
+              value: 172.16.1.100/31
+      nodeSelector:
+        kubernetes.io/hostname: ${NODES[1]}
+EOF
+```
+
+Deploy NSC and NSE:
+```bash
+kubectl apply -k .
+```
+
+Wait for applications ready:
+```bash
+kubectl wait --for=condition=ready --timeout=1m pod -l app=nsc-kernel -n ${NAMESPACE}
+```
+```bash
+kubectl wait --for=condition=ready --timeout=1m pod -l app=nse-kernel -n ${NAMESPACE}
+```
+
+Find NSC and NSE pods by labels:
+```bash
+NSC=$(kubectl get pods -l app=nsc-kernel -n ${NAMESPACE} --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
+```
+```bash
+NSE=$(kubectl get pods -l app=nse-kernel -n ${NAMESPACE} --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
+```
+
+Ping from NSC to NSE:
+```bash
+kubectl exec ${NSC} -n ${NAMESPACE} -- ping -c 4 172.16.1.100
+```
+
+Ping from NSE to NSC:
+```bash
+kubectl exec ${NSE} -n ${NAMESPACE} -- ping -c 4 172.16.1.101
+```
+
+Find local NSMgr pod:
+```bash
+NSMGR=$(kubectl get pods -l app=nsmgr --field-selector spec.nodeName==${NODES[0]} -n nsm-system --template '{{range .items}}{{.metadata.name}}{{"\n"}}{{end}}')
+```
+
+Restart local NSMgr:
+```bash
+kubectl delete pod ${NSMGR} -n nsm-system
+```
+
+Ping from NSC to NSE again after local NSMgr restored:
+```bash
+sleep 70
```

Comment:
Why do we need to use sleep?

Reply:
Heal takes 60 seconds to start, so even if we don't make any heal actions in these restore tests, we still need this wait to make sure that heal doesn't lead us to some unexpected behavior. Example:
1. NSMgr restarts.
2. We are not waiting for any heal actions - just starting with ping.
3. Ping works, because NSC, Forwarder, NSE are still working and didn't close the connection.
4. Test completes.
5. Due to some error in logic heal starts and closes the connection in Forwarder.
6. Ping is actually not working, but test doesn't check for it.

Reply:
ok, got it

Reply:
@edwarnicke Do you mind if we'll use sleep in heal tests? As I can see Vlad has created an issue to consider the problem with sleep after release https://github.com/networkservicemesh/sdk/issues/978

Reply:
> Due to some error in logic heal starts and closes the connection in Forwarder.

This sounds like a real bug we should fix.

Generally sleep is primarily a means of papering over bugs (not always, but often)... lets track down the bug instead.

Reply:
@edwarnicke The problem is tracked https://github.com/networkservicemesh/sdk/issues/978

Reply:
Should we try to consider it at first or we can go ahead and merge this PR?

Reply:
> > Due to some error in logic heal starts and closes the connection in Forwarder.
> 
> This sounds like a real bug we should fix.
> 
> Generally sleep is primarily a means of papering over bugs (not always, but often)... lets track down the bug instead.

Sorry for such confusing you, these are not some real 1..6 steps with some real bug, it is just an example of how we can possibly miss something if we will remove this sleep from tests.